logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java

```diff
@@ -599,9 +609,14 @@ public synchronized Batch nonBlockReadBatch(int limit) throws IOException {
      * @param limit size limit of the batch to read. returned {@link Batch} can be smaller.
      * @param timeout the maximum time to wait in milliseconds on write operations
      * @return the read {@link Batch} or null if no element upon timeout
+     * @throws QueueRuntimeException if queue is closed
      * @throws IOException if an IO error occurs
      */
     public synchronized Batch readBatch(int limit, long timeout) throws IOException {
```

Comment:
Deep dive note: Both `readBatch` and `nonBlockReadBatch` were _intentionally_ made `synchronized` in https://github.com/elastic/logstash/pull/9646:

> The actual behaviour change happens by making the read methods `synchronized`. **This ensures that all elements in a batch have contiguous sequence numbers.** The reason for this is, that now only a single reader goes to sleep on the `empty` `Condition` at any given time.